File: TwitchDVR/TwitchDVR/KickSession.swift
import Foundation

/// Persisted Kick session (cookie jar) captured from the in-app login WebView.
/// Recording works without it — signing in is an opt-in for ad behavior etc.
enum KickSession {
    static let storageKey = "kick_cookies"

    /// Dedicated ephemeral session so only the explicit Cookie header is sent —
    /// plus the shared URLSession cookie jar doesn't interfere.
    static let httpSession: URLSession = {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 20
        return URLSession(configuration: config)
    }()

    static var cookies: [String: String] {
        guard let raw = ConfigStore.load(key: storageKey),
              let data = raw.data(using: .utf8),
              let dict = try? JSONDecoder().decode([String: String].self, from: data) else {
            return [:]
        }
        return dict
    }

    static var isLoggedIn: Bool { !cookies.isEmpty }

    static var cookieHeader: String? {
        let parts = cookies.map { "\($0.key)=\($0.value)" }
        return parts.isEmpty ? nil : parts.joined(separator: "; ")
    }

    static func save(_ newCookies: [String: String]) {
        guard !newCookies.isEmpty else { return }
        if let data = try? JSONEncoder().encode(newCookies),
           let json = String(data: data, encoding: .utf8) {
            ConfigStore.save(key: storageKey, value: json)
        }
    }

    static func clear() {
        ConfigStore.delete(key: storageKey)
    }

    /// Fetches the signed-in username (nil when not logged in / request fails).
    static func fetchUsername() async -> String? {
        guard isLoggedIn else { return nil }
        let url = URL(string: "https://kick.com/api/v2/users/me")!
        var request = URLRequest(url: url)
        request.timeoutInterval = 15
        request.setValue(
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0 Safari/537.36",
            forHTTPHeaderField: "User-Agent")
        if let header = cookieHeader {
            request.setValue(header, forHTTPHeaderField: "Cookie")
        }
        guard let (data, response) = try? await httpSession.data(for: request),
              let http = response as? HTTPURLResponse, http.statusCode == 200,
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return obj["username"] as? String
    }
}

File: Windows/TwitchDVR.Core/KickSession.cs
using System.Text.Json;

namespace TwitchDVR.Core;

/// <summary>
/// Persisted Kick session (cookie jar) captured from the in-app login WebView.
/// Recording works without it — signing in is an opt-in for ad behavior etc.
/// </summary>
public static class KickSession
{
    public const string StorageKey = "kick_cookies";

    /// <summary>Dedicated client with cookies disabled so only the explicit Cookie header
    /// is sent and responses never populate a shared cookie jar.</summary>
    internal static readonly HttpClient Http = CreateNoCookieClient();

    private static HttpClient CreateNoCookieClient()
    {
        var handler = new HttpClientHandler { UseCookies = false };
        return new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(30) };
    }

    public static Dictionary<string, string> Cookies
    {
        get
        {
            var raw = ConfigStore.Load(StorageKey);
            if (string.IsNullOrEmpty(raw)) return new Dictionary<string, string>();
            try
            {
                return JsonSerializer.Deserialize<Dictionary<string, string>>(raw)
                       ?? new Dictionary<string, string>();
            }
            catch
            {
                return new Dictionary<string, string>();
            }
        }
    }

    public static bool IsLoggedIn => Cookies.Count > 0;

    public static string? CookieHeader
    {
        get
        {
            var parts = Cookies.Select(kv => $"{kv.Key}={kv.Value}").ToList();
            return parts.Count == 0 ? null : string.Join("; ", parts);
        }
    }

    public static void Save(Dictionary<string, string> newCookies)
    {
        if (newCookies.Count == 0) return;
        ConfigStore.Save(StorageKey, JsonSerializer.Serialize(newCookies));
    }

    public static void Clear() => ConfigStore.Delete(StorageKey);

    /// <summary>Fetches the signed-in username (null when not logged in / request fails).</summary>
    public static async Task<string?> FetchUsernameAsync()
    {
        if (!IsLoggedIn) return null;
        using var request = new HttpRequestMessage(HttpMethod.Get, "https://kick.com/api/v2/users/me");
        request.Headers.TryAddWithoutValidation("User-Agent", ChromeUserAgent);
        if (CookieHeader is { } header)
            request.Headers.TryAddWithoutValidation("Cookie", header);

        try
        {
            var response = await Http.SendAsync(request);
            if (response.StatusCode != System.Net.HttpStatusCode.OK) return null;
            var json = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.TryGetProperty("username", out var user) &&
                   user.ValueKind == JsonValueKind.String ? user.GetString() : null;
        }
        catch
        {
            return null;
        }
    }

    private const string ChromeUserAgent =
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0 Safari/537.36";
}